### PR TITLE
WT-10946 Usage patterns architecture guide

### PR DESCRIPTION
**Pull request by luke-pearson**

### CODE REVIEW

File: src/docs/arch-usage-patterns.dox [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+array pointer. This requires an @ref acq-marked. If this ordering is not respected the reader could
+read an old array pointer but a new \c slot_count and encounter an array index out of bounds
+exception.
+
+@subsubsection growable-arrays-further-details Further details
+The fixed size version of this pattern is used to maintain the connection's sessions array. The
+dynamically growing version of this pattern is used to maintain the hazard pointer array for a
+session.
+
+FIXME: add some references to the above section.
```
Comment (luke-pearson) [inline src/docs/arch-usage-patterns.dox:124]: I'll do this asap.

**Review by ajmorton — COMMENTED** [review]
I've pushed a commit with grammar nits rather than opening comment threads. Feel free to push back on any of them.

File: src/docs/arch-usage-patterns.dox [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+time the write to \c W has become visible. As a result when the \c WwS pattern succeeds, by the next
+time the value of \c S changes all threads can see the update to \c W.
+
+In this pattern the write to \c W always takes place, however the writing thread will not consider
+the update to be valid until it has verified that \c S is unchanged. The writing thread is
+responsible for either retrying or rolling back the write to \c W on a failure. Threads reading from
+\c W must be able to handle or ignore these interim "failed" writes to \c W, and must perform a
+read-acquire on \c S before reading \c W.
+
+<img src="WwS.png" alt="" style="transform: scale(0.75,0.75);" />
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/docs/arch-usage-patterns.dox:136]: [5c7b2d5](https://github.com/wiredtiger/wiredtiger/pull/10843/commits/5c7b2d595284489b1477d8e1385df8e6edf6d000)

File: src/docs/arch-usage-patterns.dox
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/docs/arch-usage-patterns.dox:1]: [adfdddb](https://github.com/wiredtiger/wiredtiger/pull/10843/commits/adfdddb2677b58cb1c7f55908654b34d7eadb385)

File: src/docs/arch-usage-patterns.dox [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+@subsection write-while-stable-pseudocode Pseudocode
+\code{.c}
+    void example_WwS(int *W, int *S) {
+        int initial_s_value = *S;
+
+        // The initial read from S must take place before the write to W, however this
+        // doesn't always require a barrier to be enforced. The function could early
+        // exit based on the value of S (__wt_hazard_set_func), or W could have a data
+        // dependency on S (__wt_session_gen_enter).
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/docs/arch-usage-patterns.dox]: [c4a4279](https://github.com/wiredtiger/wiredtiger/pull/10843/commits/c4a4279ad487d2f876ce070644d0ed40863bfacb)

File: src/include/connection.h [hunk 4]
```diff
@@ -335,8 +335,8 @@ extern const WT_NAME_FLAG __wt_stress_types[];
 
 /*
  * Access the array of all sessions. This field uses the Slotted Array pattern to managed shared
- * accesses, if you are looking to walk all sessions please consider using the existing session walk
- * logic. FIXME-WT-10946 - Add link to Slotted Array docs.
+ * accesses, if you need to walk all sessions please call __wt_session_array_walk. For more details
```
Comment (ajmorton) [inline src/include/connection.h:338]: We should still mention this is the slotted array pattern in the comment.

File: src/session/session_helper.c [hunk 5]
```diff
@@ -34,8 +34,8 @@ __wt_session_array_walk(WT_SESSION_IMPL *session,
     /*
      * Ensure we read the session count only once. We want to iterate over all sessions that were
      * active at this point in time. Sessions in the array may open, close, or be have their
-     * contents change during traversal. We expect the calling code to handle this. See the slotted
-     * sessions docs for further details. FIXME-WT-10946 Add link to docs once they're added.
+     * contents change during traversal. We expect the calling code to handle this. See the usage
```
Comment (ajmorton) [inline src/session/session_helper.c:37]: Same here about referencing slotted arrays

File: src/docs/arch-usage-patterns.dox [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+time the write to \c W has become visible. As a result when the \c WwS pattern succeeds, the next
+time the value of \c S changes all threads can already see the update to \c W.
+
+In this pattern the write to \c W always takes place, however the writing thread will not consider
+the update to be valid until it has verified that \c S is unchanged. The writing thread is
+responsible for either retrying or rolling back the write to \c W on a failure. Threads reading from
+\c W must be able to handle or ignore these interim "failed" writes to \c W, and must perform a
+read-acquire on \c S before reading \c W.
+
+<img src="WwS.png" alt="" style="transform: scale(0.75,0.75);" />
```
Comment (ajmorton) [inline src/docs/arch-usage-patterns.dox:136]: nit: This image is rather large in the generated docs. 
I tried changing scale to (0.5, 0.5) but it still looked ugly so happy to leave as is if we can't find a quick fix.

**Review by ajmorton — APPROVED** [review]
LGTM. Some nits but nothing that needs re-review

File: src/include/connection.h (hunk 4 above)
Comment (luke-pearson) [reply src/include/connection.h:338]: It does lol.

File: src/docs/arch-usage-patterns.dox (hunk 6 above)
Comment (luke-pearson) [reply src/docs/arch-usage-patterns.dox:136]: Yeah, I'm not sure how well the scale config works but I'll leave it as is for now.

File: src/include/connection.h (hunk 4 above)
Comment (ajmorton) [reply src/include/connection.h:338]: So it does. I got fixated on the red diff

**Pull request merged**